File: imperium-common/src/main/java/com/xpdustry/imperium/common/string/StringTrieMap.java
package com.xpdustry.imperium.common.string;

import java.util.List;
import org.jspecify.annotations.Nullable;

public interface StringTrieMap<V> {

    static <V> StringTrieMap.Mutable<V> create() {
        return new StringTrieMapImpl<>();
    }

    List<Token<V>> search(final CharSequence chars);

    @Nullable V get(final CharSequence chars);

    boolean contains(final CharSequence chars, final boolean partial);

    record Token<V>(String word, int index, V value) {}

    interface Mutable<V> extends StringTrieMap<V> {

        @Nullable V put(final CharSequence chars, final V value);
    }
}

File: imperium-common/src/main/kotlin/com/xpdustry/imperium/common/string/StringTrieMap.kt
// SPDX-License-Identifier: GPL-3.0-only
package com.xpdustry.imperium.common.string

interface StringTrieMap<V> {

    fun search(chars: CharSequence): List<Token<V>>

    operator fun get(chars: CharSequence): V?

    fun contains(chars: CharSequence, partial: Boolean): Boolean

    data class Token<V>(val word: String, val index: Int, val value: V)

    interface Mutable<V> : StringTrieMap<V> {
        fun put(chars: CharSequence, value: V): V?
    }

    companion object {
        fun <V> create(): Mutable<V> = StringTrieMapImpl()
    }
}
